zebra-consensus/src/parameters/network_upgrade.rs

```diff
@@ -0,0 +1,163 @@
+//! Network upgrade consensus parameters for Zcash.
+
+use NetworkUpgrade::*;
+
+use std::collections::{BTreeMap, HashMap};
+use std::ops::Bound::*;
+
+use zebra_chain::types::BlockHeight;
+use zebra_chain::{Network, Network::*};
+
+/// A Zcash network protocol upgrade.
+//
+// TODO: are new network upgrades a breaking change, or should we make this
+//       enum non-exhaustive?
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum NetworkUpgrade {
+    /// The Zcash protocol before the Overwinter upgrade.
+    ///
+    /// We avoid using `Sprout`, because the specification says that Sprout
+    /// is the name of the pre-Sapling protocol, before and after Overwinter.
+    BeforeOverwinter,
+    /// The Zcash protocol after the Overwinter upgrade.
+    Overwinter,
+    /// The Zcash protocol after the Sapling upgrade.
+    Sapling,
+    /// The Zcash protocol after the Blossom upgrade.
+    Blossom,
+    /// The Zcash protocol after the Heartwood upgrade.
+    Heartwood,
+    /// The Zcash protocol after the Canopy upgrade.
+    Canopy,
+}
+
+/// Mainnet network upgrade activation heights.
+///
+/// This is actually a bijective map, but it is const, so we use a vector, and
+/// do the uniqueness check in the unit tests.
+pub(crate) const MAINNET_ACTIVATION_HEIGHTS: &[(BlockHeight, NetworkUpgrade)] = &[
+    (BlockHeight(0), BeforeOverwinter),
+    (BlockHeight(347_500), Overwinter),
+    (BlockHeight(419_200), Sapling),
+    (BlockHeight(653_600), Blossom),
+    (BlockHeight(903_000), Heartwood),
+    (BlockHeight(1_046_400), Canopy),
+];
+
+/// Testnet network upgrade activation heights.
+///
+/// This is actually a bijective map, but it is const, so we use a vector, and
+/// do the uniqueness check in the unit tests.
+pub(crate) const TESTNET_ACTIVATION_HEIGHTS: &[(BlockHeight, NetworkUpgrade)] = &[
+    (BlockHeight(0), BeforeOverwinter),
+    (BlockHeight(207_500), Overwinter),
+    (BlockHeight(280_000), Sapling),
+    (BlockHeight(584_000), Blossom),
+    (BlockHeight(903_800), Heartwood),
+    // As of 21 July 2020, the Canopy testnet height has not been decided.
+    // See ZIP 251 for updates.
+];
+
+/// The Consensus Branch Id, used to bind transactions and blocks to a
+/// particular network upgrade.
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub struct ConsensusBranchId(u32);
+
+/// Network Upgrade Consensus Branch Ids.
+///
+/// Branch ids are the same for mainnet and testnet. If there is a testnet
+/// rollback after a bug, the branch id changes.
+///
+/// Branch ids were introduced in the Overwinter upgrade, so there is no
+/// BeforeOverwinter branch id.
+///
+/// This is actually a bijective map, but it is const, so we use a vector, and
+/// do the uniqueness check in the unit tests.
+pub(crate) const CONSENSUS_BRANCH_IDS: &[(NetworkUpgrade, ConsensusBranchId)] = &[
+    // TODO(teor): byte order?
```

Comment:
@dconnolly can you check I have the right byte order here?

I used the constants from the ZIPs:
* https://zips.z.cash/zip-0201#deployment-of-overwinter
* https://zips.z.cash/zip-0205#sapling-deployment
* https://zips.z.cash/zip-0206#blossom-deployment
* https://zips.z.cash/zip-0250#heartwood-deployment
* https://zips.z.cash/zip-0251#canopy-deployment